LUCENE-9322: add VectorValues to new Lucene90 codec

msokolov opened the pull request:
This adds a floating-point vector format building on the designs in Lucene-9004 and LUCENE-9322. This patch fully supports indexing and reading vectors with an iterator, and a random-access API. Support for search based on an NSW graph implementation is intended to follow soon, but I wanted to include the vector APIs that I needed to get that working, even though they are not yet used here, so eg it includes the definition of a scoring function and a nearest-neighbors search API, but no implementation of search yet. My intention is to keep the ANN implementation hidden, so graphs and other supporting data structures (eg we might want to support LSH or k-means clustering and so on) would be implementation details invoked by a configuration on the VectorField/VectorValues. At the moment you can specify a ScoringFunction, and it is implicit that NSW will be the result. In the future we could add another parameter to ScoringFunction and/or new functions to represent support for other algorithms.

Some open questions: 

1. Should this be Lucene 9.0 only? In this patch I added Lucene90 Codec. If we do this then it would be awkward to backport.
2. It seems messy to have the ScoringFunction implementation in the main VectorValues interface API file. I'd appreciate any better suggestion for how to organize this.
3. Vector scoring can return negative numbers. I'd like to have first-class support for dot product distance (which can be negative) since that's what my consumers seem to have settled on. I don't think we need to be compatible with relevance scores, at least not directly in the KNN search API, but IDK maybe we should? We could renormalize/convert from dot-product scores to a positive score with math in the output layer where we return the scores. So far this is just specification question as there is no implementation of search yet.
4. I think there is room for improvement in some of the data structures used to map docids to dense vector ordinals and back. I'd appreciate comments on that, but maybe we could revisit in a fast follow-on issue?

mikemccand reviewed (commented):
This is impressive @msokolov!  Not every day we add a new `Codec` format!  It looks quite mature already... I left a bunch of small comments.

Thank you for the extensive test cases, especially the `testRandom`, and the tests catching mis-use where user tries to change dimension or scoring function in an existing field.

I would love to see a "Vector Overview" javadoc somewhere for users to understand what Lucene can do with vectors.  E.g., state that you add a `VectorField` to your document, they are single valued, always type `float`, always dense.  They are vectors (1 "axis", i.e. not matrices).  Dimensionality and score function cannot change once the first doc is indexed.  The scoring must be set at index-time because in the future, Lucene must do index-time work to set index up to efficiently execute that scoring at search time (?).  At search time you can retrieve vectors by iterator or random access, and apply the pairwise score function.  The new default codec (Lucene90) efficiently handles sparse case (some docs have vectors and some do not) but otherwise does not do any exotic vector compression beyond that (which is great for first cut!).

I see you implemented the two score functions,  but are they ever exercised in tests?  I see the static  `ScoreFunction.compare` method but it looks uninvoked.

> Should this be Lucene 9.0 only? In this patch I added Lucene90 Codec. If we do this then it would be awkward to backport.

+1 for 9.0 only.

> I think there is room for improvement in some of the data structures used to map docids to dense vector ordinals and back. I'd appreciate comments on that, but maybe we could revisit in a fast follow-on issue?

+1 to get the functionality working first and optimize later.

I am curious how the basic vector usage performs -- just indexing one vector field, and retrieving it at search time.  We can (separately) enable `luceneutil` to support testing vectors, somehow.  But I wonder where we'll get semi-realistic vectors derived from Wikipedia content ...

msokolov commented:
Thanks for the extensive comments, @mikemccand - I'll address soon with an updated PR. 

I also found some bugs in implementations of the random access interface, and I want to fix those and enhance the test coverage. Mainly, they were incorrectly sharing state with the enclosing iterators, which causes issues if you simultaneously iterate and access randomly. There were also a couple of bugs dealing with empty segments that apparently weren't caught by testRandom.

msokolov commented:
> Thank you for ... the tests catching mis-use where user tries to change dimension or scoring function in an existing field.
Thanks to @mocobeta for those; I was able to carry that forward from her earlier patch

> I see you implemented the two score functions, but are they ever exercised in tests
True - this was extracted from a bigger change including usage of those methods as part of KNN search, but they deserve their own unit tests - I'll add.

> I would love to see a "Vector Overview" javadoc somewhere ...
Yes - I'll add to the VectorValues/VectorField class javadocs I think that's the most natural/visible place.

> I am curious how the basic vector usage performs -- just indexing one vector field, and retrieving it at search time. We can (separately) enable luceneutil to support testing vectors, somehow. But I wonder where we'll get semi-realistic vectors derived from Wikipedia content 
Agreed that benchmarking is needed. I think we can use http://ann-benchmarks.com/ as a guide for some standardized test vectors. They won't be related to wikipedia? If we get to wanting that, we could also make use of something like https://fasttext.cc/docs/en/pretrained-vectors.html that is trained on ngrams taken from Wikipedia (for many languages)? I don't know how suited it is, just found in a google search. For that, we'd have to compute document/query vectors based on an ngram-vector dictionary. I think a simple thing is to sum all the ngram-vectors for all the ngrams in a document / query

msokolov commented on lucene/core/src/java/org/apache/lucene/codecs/VectorWriter.java:
```diff
@@ -0,0 +1,253 @@
+          } else if (dimension != segmentDimension) {
+            throw new IllegalStateException("Varying dimensions for vector-valued field " + mergeFieldInfo.name
+                + ": " + dimension + "!=" + segmentDimension);
+          } else if (scoreFunction != segmentScoreFunction) {
```
Yes, IndexingChain.indexVector calls FieldInfo.setVectorDimensionAndScoreFunction, which checks against existing values. It allows to go from 0 to non-zero dimension (setting scoreFunction at that time), but no other change is allowed. This is tested in TestVectorValues

msokolov commented on lucene/core/src/java/org/apache/lucene/document/FieldType.java:
```diff
@@ -351,6 +356,27 @@ public int pointNumBytes() {
+    if (numDimensions > VectorValues.MAX_DIMENSIONS) {
+      throw new IllegalArgumentException("vector numDimensions must be <= VectorValues.MAX_DIMENSIONS (=" + VectorValues.MAX_DIMENSIONS + "); got " + numDimensions);
+    }
+    this.vectorDimension = numDimensions;
```
hmm I noticed we do not do this for Points. I think that in practice we do not expect users to create these FieldTypes - they are created implicitly when adding values using VectorField constructors - we create a new Type for every Field unless the user cleverly creates a generic Field with a float[] as its `fieldsData` and an appropriate FieldType. Efficient usage would seem to be: create a VectorField, add it to a doc, update its value and repeat. Anyway I'll add the checking

msokolov commented on lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:
```diff
@@ -374,7 +377,25 @@ private FieldNormStatus() {
       public int totalValueFields;
       
-      /** Exception thrown during doc values test (null on success) */
+      /** Exception thrown during point values test (null on success) */
```
Yeah, Points was a model for much of this, so I stumbled on that

msokolov commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorReader.java:
```diff
@@ -0,0 +1,326 @@
+
+    @Override
+    public int advance(int target) throws IOException {
+      // We could do better by log-binary search in ordToDoc, but this is never used
```
Yes, the doc-> ord relation is monotonic, ie order-preserving s.t. `doc1 > doc2 iff ord(doc1) > ord(doc2)`. We use this fact in the implementation of random-access over a sorted index while merging so I don't think we're free to share vectors like that (with this implementation). I also plan to make use of that invariant when building NSW graphs - we want to store the ords in sorted order, and if we generate the graph while indexing, then they will naturally be added in order (since their order corresponds to the docid order) with no need to sort them at any point.

I tend to think it's unlikely that you would often see duplicated floating-point vectors aside from zeros, although of course YMMV. I think we will want to allow for lossy compression using byte-per-dimension rather than float-per-dimension, but I am hard-pressed to see what kind of lossless compression we could provide. Well I see there is stuff out there: https://computing.llnl.gov/projects/floating-point-compression , but I expect it would require some additional index structure to enable random access

msokolov commented on lucene/codecs/src/java/org/apache/lucene/codecs/simpletext/SimpleTextVectorReader.java:
```diff
@@ -0,0 +1,304 @@
+    }
+
+    @Override
+    public BytesRef binaryValue() {
```
Yeah, seems a bit weird from the API standpoint, yet (1) we primarily want to expose `float[] vectorValue` yet for efficiency when flushing and merging we don't care what the `float[]` value is and just want access to the underlying bytes. Maybe there is a way to provide some private interface for IndexWriter, but I didn't see how.

msokolov commented on lucene/core/src/java/org/apache/lucene/index/VectorValuesWriter.java:
```diff
@@ -0,0 +1,272 @@
+    }
+    assert docID > lastDocID;
+    docsWithField.add(docID);
+    vectors.add(ArrayUtil.copyOfSubArray(vectorValue, 0, vectorValue.length));
```
Yes in the normal case, but a sneaky/unlucky person could bypass Field type safety; I'll add a check here.

msokolov commented on lucene/core/src/java/org/apache/lucene/index/VectorValues.java:
```diff
@@ -0,0 +1,264 @@
+   * Score values may be negative. By default high scores indicate nearer documents, unless the function is reversed.
+   */
+  public enum ScoreFunction {
+    /** No distance function is used. Note: {@link VectorValues.RandomAccess#search(float[], int, int)}
```
OK I opened LUCENE-9573. I have to admit I don't fully understand the timing constraints/dependencies here. Maybe you could comment on that issue?  Re: the ids I opted to move to using the enum ordinal as you suggested later. I can't see how that restricts us in any meaningful way. Perhaps we add a back-compat test to verify that the enum ordinals don't change

jtibshirani commented:
> Hmm I tried to get that benchmarking suite to run and it requires some major Python-fu.

I managed to get this working a few months ago while experimenting with a clustering-based approach: https://github.com/jtibshirani/ann-benchmarks/pull/2. It indeed involved a lot of set-up -- I can try to get it working again post results. Going forward, I think it will be helpful to use ann-benchmarks to compare recall and QPS against the ANN reference implementations.

jtibshirani commented:
>  It seems to expect your algorithm to be delivered as an in-process extension to Python, which works OK for a native code library, but I'm not sure how we'd present Lucene to it. We don't want to have to call through a network API?

I ended up using `py4j` to call out to Lucene, which sets up a 'gateway server' and passes data between the Python + Java processes through a socket. I did find there to be a significant overhead just from converting between Python <-> Java, but can be largely mitigated by making sure to use 'batch mode' (the `--batch` option), which allows all query vectors to be passed to Lucene at once. Amortizing the overhead this way, I was able to get consistent + informative results. Let me know if you're interested in trying the py4j option and I can post set-up steps. I found it helpful while developing but it's quite tricky and maybe shouldn't be the main way to track performance right now (as you mentioned) !

A note that it's possible to use vector data from ann-benchmarks without integrating with the framework. The datasets are listed [here](https://github.com/erikbern/ann-benchmarks/blob/master/ann_benchmarks/datasets.py#L396) and made available on the website in hdf5 format. One option could be 200-dimensional GloVe word vectors, available from `http://ann-benchmarks.com/glove-200-angular.hdf5`. I think these are trained on Twitter data.

msokolov commented:
> One option could be 200-dimensional GloVe word vectors, available from http://ann-benchmarks.com/glove-200-angular.hdf5. I think these are trained on Twitter data.

+1 I'm looking into adding GloVe data to luceneutil benchmarks, initially just to index and retrieve them, then I hope to add tasks for scoring lexical matches, and then for knn matching. I think some of the GloVe datasets are trained on wikipedia (plus other text) so should be suitable for use in our benchmarks, which are based on wikipedia text.

I think for initial performance comparisons we can use our own tool; it wouldn't be as nicely controlled as running in the same framework, but if we are careful the results should be comparable. And it's good to know there is a reasonable path for integrating with ann-benchmark using py4j, and I hadn't realized there was a --batch option.

msokolov commented:
So .. this has been out for 17 days now - I'd like to commit in the next day or so in order to enable moving forward with the matching piece of this, adding some benchmarking to luceneutil that depends on it, etc. @mikemccand do you want to have another look? - I think I've addressed your comments.

mikemccand commented:
> So .. this has been out for 17 days now - I'd like to commit in the next day or so in order to enable moving forward with the matching piece of this, adding some benchmarking to luceneutil that depends on it, etc. @mikemccand do you want to have another look? - I think I've addressed your comments.

Thanks @msokolov -- I'll have another look today.  It was already close the first time I looked :)

mikemccand commented on lucene/core/src/test/org/apache/lucene/index/TestVectorValues.java:
```diff
@@ -0,0 +1,650 @@
+    }
+  }
+
+  public void testIllegalScoreFunctionChangeViaAddIndexesSlowCodecReader() throws Exception {
```
We have good coverage of illegal `addIndexes` usage, but do we have a test case of a successful usage, adding in an index with vectors to an index without, and vice/versa, and then both?

msokolov replied:
Ah, good idea, I will add

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/VectorValues.java:
```diff
@@ -0,0 +1,273 @@
+   * The returned array may be shared across calls, re-used, and modified as the iterator advances.
+   * @return the vector value
+   */
+  public abstract float[] vectorValue() throws IOException;
```
Hmm, is there an API somewhere to let me get the vector ordinal for the current Lucene `docId`?

msokolov replied:
If you are iterating, you can keep track by counting. On the other hand, you wouldn't need to since the only purpose for the ordinal is to retrieve the vector that you already have. So I think getting the ordinal for  the *current* docId is not really all that helpful.

In theory one might want to get an ordinal for some arbitrary docId. I guess the API is sort of incomplete without it - it offers random access, but only by an opaque ordinal. You can of course iterate over all the docs and build your own map, but that is kind of unhelpful.

However supporting this comes at some additional cost. It's not required to support knnSearch since the way that works is to search for the best ordinal and then map that to a docid, but internally we do not maintain any docid->ordinal mapping. We can get an answer using binary search in the ordToDoc map, but I wonder if we should expose that. WDYT?

msokolov replied:
Thinking about this some more: why do we have the random access interface? It enables a vector consumer to build efficient data structures using vector ordinals as keys, allowing them to avoid an extra step of mapping from docid to vector ordinal and back. It's expected that such a consumer will maintain their own such mapping - the point is that we don't want to force them to constantly be mapping (by eg defining this API in terms of docids). And we don't want to provide a less-than-ideal mapping function, or spend extra effort to maintain a mapping that may not be used.

We could try hiding this behind a more opaque class structure, but given that this interface needs to be accessed in o.a.l.index and o.a.l.codecs I think it would have to be public, and I'm not sure what the benefit would be.

I think we can always *add* `ordinal(int docId)` and `docId(int ordinal)` methods as they become useful? I think that would be a good API, and nothing here prevents it from being added later, so I propose to commit this as is.

mikemccand reviewed (approved):
The change looks great!  Vectors are born (in Lucene)!

I left minor comments -- let's make sure to add `@lucene.experimenal` to the new vector classes to reserve for future freedom.

> Thanks to @mocobeta for those; I was able to carry that forward from her earlier patch

+1, thank you @mocobeta -- this is a great collaborative effort!

msokolov commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorReader.java:
```diff
@@ -0,0 +1,344 @@
+
+/**
+ * Reads vectors from the index segments.
+ */
```
ok; tagging experimental here and the other classes you mentioned. All the public classes I guess? Should we also tag o.a.l.codecs.VectorReader/Writer and o.a.l.document/VectorField

msokolov commented on lucene/core/src/java/org/apache/lucene/index/VectorValues.java:
```diff
@@ -0,0 +1,273 @@
+   * The returned storage may be shared across calls, re-used and modified as the iterator advances.
+   * @return the binary value
+   */
+  public BytesRef binaryValue() throws IOException {
```
Yes, I'll remove the mention of the precise encoding.

msokolov commented:
I squashed, resolved conflicts and pushed using command line

jtibshirani commented:
@msokolov I did a belated review and had some high-level thoughts. Hopefully post-commit reviews are still helpful in giving ideas for follow-ups.

> It seems messy to have the ScoringFunction implementation in the main VectorValues interface API file. 

It seems like we’re using `ScoreFunction` for two separate use cases:
1. Configuring the ANN indexing + search strategy (to be introduced later). For example, HNSW needs to know up-front what similarity measure to use when building the graph.
2. Providing convenience methods for users that want to incorporate vectors into scoring, outside of ANN.

We could think about separating out these cases. Maybe `ScoreFunction` could stick closely to case 1 and avoid exposing score implementations. This fits well with its role of specifying the ANN index strategy, which might encompass other ANN approaches in the future like `KMEANS_EUCLIDEAN`. And for use case 2, we could introduce static utility methods like `dotProduct(float[] a, float[] b)` ? The ANN implementations would be free to make use of these methods if they prove helpful. If we go with this design, perhaps `ScoreFunction` would be renamed to something like `SearchStrategy`.

Separately, I wonder if we need to expose a random access interface on the public `VectorValues`. When using vectors for scoring, iterator access seems sufficient. I’m having trouble thinking of a use for random access outside of ANN, and some ANN implementations don’t even need it -- for example most k-means and LSH approaches only need forward iteration. Would it make sense to keep the `VectorValues` surface area smaller and only provide iterator access? Random access could still be supported at a lower level, in a place that's available to HNSW. I’m not sure exactly how this would look, perhaps it'd be clearer once we see how HNSW will be integrated.

jtibshirani reviewed (commented):
@msokolov I did a belated review and had some high-level thoughts. Hopefully post-commit reviews are still helpful in giving ideas for follow-ups.

It seems messy to have the ScoringFunction implementation in the main VectorValues interface API file.

It seems like we’re using ScoreFunction for two separate use cases:

Configuring the ANN indexing + search strategy (to be introduced later). For example, HNSW needs to know up-front what similarity measure to use when building the graph.
Providing convenience methods for users that want to incorporate vectors into scoring, outside of ANN.
We could think about separating out these cases. Maybe ScoreFunction could stick closely to case 1 and avoid exposing score implementations. This fits well with its role of specifying the ANN index strategy, which might encompass other ANN approaches in the future like KMEANS_EUCLIDEAN. And for use case 2, we could introduce static utility methods like dotProduct(float[] a, float[] b) ? The ANN implementations would be free to make use of these methods if they prove helpful. If we go with this design, perhaps ScoreFunction would be renamed to something like SearchStrategy.

Separately, I wonder if we need to expose a random access interface on the public VectorValues. When using vectors for scoring, iterator access seems sufficient. I’m having trouble thinking of a use for random access outside of ANN, and some ANN implementations don’t even need it -- for example most k-means and LSH approaches only need forward iteration. Would it make sense to keep the VectorValues surface area smaller and only provide iterator access? Random access could still be supported at a lower level, in a place that's available to HNSW. I’m not sure exactly how this would look, perhaps it'd be clearer once we see how HNSW will be integrated.

msokolov commented:
@jtibshirani I am in sympathy with your points; the scoring implementations don't really belong in VectorValues which is more of an abstract interface otherwise, and random-access API should really be an internal detail of the HNSW implementation. I did fiddle with ideas for hiding the RandomAccess API but see my comments about visibility in the thread above; I'm not yet sure what the best way is. Probably it will be easiest to address these as part of the HNSW patch, as you say, but if you have specific ideas, please feel free to post a PR!
